File: services/surgeAlertService.js
const moment = require('moment');
const { Handle, Reply, SurgeAlertConfig, SurgeAlert } = require('../models');
const { Op } = require('sequelize');
const winston = require('winston');

const logger = winston.createLogger({
    level: 'info',
    format: winston.format.json(),
    transports: [new winston.transports.Console()]
});

class SurgeAlertService {
    static async getThroughputMetrics(handleId) {
        const now = moment();
        const fifteenMinutesAgo = moment().subtract(15, 'minutes');
        const hourAgo = moment().subtract(1, 'hour');
        const dayAgo = moment().subtract(24, 'hours');

        try {
            const [fifteenMinCount, hourCount, dayCount] = await Promise.all([
                Reply.count({
                    where: {
                        HandleId: handleId,
                        isHidden: true,
                        hiddenAt: {
                            [Op.gte]: fifteenMinutesAgo.toDate()
                        }
                    }
                }),
                Reply.count({
                    where: {
                        HandleId: handleId,
                        isHidden: true,
                        hiddenAt: {
                            [Op.gte]: hourAgo.toDate()
                        }
                    }
                }),
                Reply.count({
                    where: {
                        HandleId: handleId,
                        isHidden: true,
                        hiddenAt: {
                            [Op.gte]: dayAgo.toDate()
                        }
                    }
                })
            ]);

            return {
                hidden_replies_last_15_min: fifteenMinCount,
                hidden_replies_last_hour: hourCount,
                hidden_replies_last_24_hours: dayCount
            };
        } catch (error) {
            logger.error('Error getting throughput metrics:', error);
            throw error;
        }
    }

    static async createConfig(handleId, data, createdBy) {
        try {
            const config = await SurgeAlertConfig.create({
                HandleId: handleId,
                surgeReplyCountPerPeriod: data.surge_reply_count_per_period,
                surgeReplyPeriodInMs: data.surge_reply_period_in_ms,
                alertCooldownPeriodInMs: data.alert_cooldown_period_in_ms || 900000, // Default 15 minutes
                emailsToNotify: data.emails_to_notify,
                enabled: data.enabled ?? true,
                createdBy: createdBy
            });

            return config;
        } catch (error) {
            logger.error('Error creating surge alert config:', error);
            throw error;
        }
    }

    static async updateConfig(configId, data, updatedBy) {
        try {
            const config = await SurgeAlertConfig.findByPk(configId);
            if (!config) {
                return null;
            }

            await config.update({
                surgeReplyCountPerPeriod: data.surge_reply_count_per_period ?? config.surgeReplyCountPerPeriod,
                surgeReplyPeriodInMs: data.surge_reply_period_in_ms ?? config.surgeReplyPeriodInMs,
                alertCooldownPeriodInMs: data.alert_cooldown_period_in_ms ?? config.alertCooldownPeriodInMs,
                emailsToNotify: data.emails_to_notify ?? config.emailsToNotify,
                enabled: data.enabled ?? config.enabled,
                updatedBy: updatedBy
            });

            return config;
        } catch (error) {
            logger.error('Error updating surge alert config:', error);
            throw error;
        }
    }

    static async getConfigs(handleId) {
        try {
            return await SurgeAlertConfig.findAll({
                where: { HandleId: handleId }
            });
        } catch (error) {
            logger.error('Error getting surge alert configs:', error);
            throw error;
        }
    }

    static async evaluateSurges() {
        try {
            logger.info('Starting surge evaluation...');
            const configs = await SurgeAlertConfig.findAll({
                where: { enabled: true },
                include: [Handle]
            });

            logger.info(`Found ${configs.length} active configurations to evaluate`);
            const now = moment();
            const results = [];

            for (const config of configs) {
                logger.info(`Evaluating config ${config.id} for handle ${config.Handle?.handle}`);
                const periodStart = moment().subtract(config.surgeReplyPeriodInMs, 'milliseconds');

                // Count hidden replies in the period
                const hiddenCount = await Reply.count({
                    where: {
                        HandleId: config.HandleId,
                        isHidden: true,
                        hiddenAt: {
                            [Op.gte]: periodStart.toDate()
                        }
                    }
                });

                logger.info(`Found ${hiddenCount} hidden replies in the last ${config.surgeReplyPeriodInMs}ms (threshold: ${config.surgeReplyCountPerPeriod})`);

                if (hiddenCount >= config.surgeReplyCountPerPeriod) {
                    logger.info('Threshold exceeded, checking cooldown period...');
                    // Check cooldown period
                    const lastAlert = await SurgeAlert.findOne({
                        where: { configId: config.id },
                        order: [['createdAt', 'DESC']]
                    });

                    let cooldownPassed = true;
                    if (lastAlert && config.alertCooldownPeriodInMs) {
                        const cooldownEnd = moment(lastAlert.createdAt)
                            .add(config.alertCooldownPeriodInMs, 'milliseconds');
                        cooldownPassed = now.isAfter(cooldownEnd);
                        logger.info(`Cooldown period ${cooldownPassed ? 'has passed' : 'is still active'}`);
                    }

                    if (cooldownPassed) {
                        logger.info('Creating new surge alert...');
                        // Create new surge alert
                        const alert = await SurgeAlert.create({
                            configId: config.id,
                            surgeAmount: hiddenCount,
                            configSnapshot: {
                                surge_reply_count_per_period: config.surgeReplyCountPerPeriod,
                                surge_reply_period_in_ms: config.surgeReplyPeriodInMs,
                                alert_cooldown_period_in_ms: config.alertCooldownPeriodInMs,
                                emails_to_notify: config.emailsToNotify
                            }
                        });
                        results.push(alert);
                        logger.info(`Created surge alert with ID ${alert.id}`);
                    }
                }

                // Update last evaluated timestamp
                await config.update({ lastEvaluatedAt: now.toDate() });
            }

            logger.info(`Surge evaluation complete. Created ${results.length} new alerts.`);
            return results;
        } catch (error) {
            logger.error('Error evaluating surges:', error);
            throw error;
        }
    }

    static async getPendingAlerts() {
        try {
            return await SurgeAlert.findAll({
                where: { alertedAt: null },
                include: [{
                    model: SurgeAlertConfig,
                    include: [Handle]
                }]
            });
        } catch (error) {
            logger.error('Error getting pending alerts:', error);
            throw error;
        }
    }

    static async markAlertSent(alertId) {
        try {
            const alert = await SurgeAlert.findByPk(alertId);
            if (alert) {
                await alert.update({ alertedAt: moment().toDate() });
            }
        } catch (error) {
            logger.error('Error marking alert as sent:', error);
            throw error;
        }
    }
}

module.exports = SurgeAlertService;

File: surge_alert_service.py
from datetime import datetime, timedelta
from typing import Dict, List, Optional
import json
from sqlalchemy import text
from app import db
from models import Handle, Reply, SurgeAlertConfig, SurgeAlert

class SurgeAlertService:
    @staticmethod
    def get_throughput_metrics(handle_id: int) -> Dict:
        """Get hidden reply throughput metrics for the specified time periods."""
        now = datetime.utcnow()
        
        # Calculate timestamps for different periods
        fifteen_min_ago = now - timedelta(minutes=15)
        hour_ago = now - timedelta(hours=1)
        day_ago = now - timedelta(days=1)
        
        # Query hidden replies for different time periods
        fifteen_min_count = Reply.query.filter(
            Reply.handle_id == handle_id,
            Reply.is_hidden == True,
            Reply.hidden_at >= fifteen_min_ago
        ).count()
        
        hour_count = Reply.query.filter(
            Reply.handle_id == handle_id,
            Reply.is_hidden == True,
            Reply.hidden_at >= hour_ago
        ).count()
        
        day_count = Reply.query.filter(
            Reply.handle_id == handle_id,
            Reply.is_hidden == True,
            Reply.hidden_at >= day_ago
        ).count()
        
        return {
            "hidden_replies_last_15_min": fifteen_min_count,
            "hidden_replies_last_hour": hour_count,
            "hidden_replies_last_24_hours": day_count
        }

    @staticmethod
    def create_config(handle_id: int, data: Dict, created_by: str) -> SurgeAlertConfig:
        """Create a new surge alert configuration."""
        config = SurgeAlertConfig(
            handle_id=handle_id,
            surge_reply_count_per_period=data['surge_reply_count_per_period'],
            surge_reply_period_in_ms=data['surge_reply_period_in_ms'],
            alert_cooldown_period_in_ms=data.get('alert_cooldown_period_in_ms', 900000),  # Default 15 minutes
            emails_to_notify=data['emails_to_notify'],
            enabled=data.get('enabled', True),
            created_by=created_by
        )
        db.session.add(config)
        db.session.commit()
        return config

    @staticmethod
    def update_config(config_id: str, data: Dict, updated_by: str) -> Optional[SurgeAlertConfig]:
        """Update an existing surge alert configuration."""
        config = SurgeAlertConfig.query.get(config_id)
        if not config:
            return None
            
        config.surge_reply_count_per_period = data.get('surge_reply_count_per_period', config.surge_reply_count_per_period)
        config.surge_reply_period_in_ms = data.get('surge_reply_period_in_ms', config.surge_reply_period_in_ms)
        config.alert_cooldown_period_in_ms = data.get('alert_cooldown_period_in_ms', config.alert_cooldown_period_in_ms)
        config.emails_to_notify = data.get('emails_to_notify', config.emails_to_notify)
        config.enabled = data.get('enabled', config.enabled)
        config.updated_by = updated_by
        
        db.session.commit()
        return config

    @staticmethod
    def get_configs(handle_id: int) -> List[SurgeAlertConfig]:
        """Get all surge alert configurations for a handle."""
        return SurgeAlertConfig.query.filter_by(handle_id=handle_id).all()

    @staticmethod
    def evaluate_surges() -> None:
        """
        Evaluate all active configurations for potential surges.
        This is called by the database event every 5 minutes.
        """
        configs = SurgeAlertConfig.query.filter_by(enabled=True).all()
        now = datetime.utcnow()
        
        for config in configs:
            # Convert period from milliseconds to seconds for timedelta
            period_seconds = config.surge_reply_period_in_ms / 1000
            period_start = now - timedelta(seconds=period_seconds)
            
            # Count hidden replies in the period
            hidden_count = Reply.query.filter(
                Reply.handle_id == config.handle_id,
                Reply.is_hidden == True,
                Reply.hidden_at >= period_start
            ).count()
            
            if hidden_count >= config.surge_reply_count_per_period:
                # Check if we're still in cooldown from last alert
                last_alert = SurgeAlert.query.filter_by(config_id=config.id)\
                    .order_by(SurgeAlert.created_at.desc()).first()
                
                cooldown_passed = True
                if last_alert and config.alert_cooldown_period_in_ms:
                    cooldown_seconds = config.alert_cooldown_period_in_ms / 1000
                    cooldown_end = last_alert.created_at + timedelta(seconds=cooldown_seconds)
                    cooldown_passed = now >= cooldown_end
                
                if cooldown_passed:
                    # Create new surge alert
                    alert = SurgeAlert(
                        config_id=config.id,
                        surge_amount=hidden_count,
                        config_snapshot=json.dumps({
                            'surge_reply_count_per_period': config.surge_reply_count_per_period,
                            'surge_reply_period_in_ms': config.surge_reply_period_in_ms,
                            'alert_cooldown_period_in_ms': config.alert_cooldown_period_in_ms,
                            'emails_to_notify': config.emails_to_notify
                        })
                    )
                    db.session.add(alert)
            
            # Update last evaluated timestamp
            config.last_evaluated_at = now
            
        db.session.commit()

    @staticmethod
    def get_pending_alerts() -> List[SurgeAlert]:
        """Get all unprocessed alerts (where alerted_at is NULL)."""
        return SurgeAlert.query.filter_by(alerted_at=None).all()

    @staticmethod
    def mark_alert_sent(alert_id: str) -> None:
        """Mark an alert as sent by updating its alerted_at timestamp."""
        alert = SurgeAlert.query.get(alert_id)
        if alert:
            alert.alerted_at = datetime.utcnow()
            db.session.commit()
